File: py/Apple/Arrays_and_Strings/myatoi.py
# convert string to integer 

def myatoi(s):
    i = 0
    numberset={"0","1","2","3","4","5","6","7","8","9"}
    # skip space 
    while (i<len(s) and s[i]==" "):
        i+=1

    # pos flag == 0 / neg flag == 1
    flag = 0
    if i<len(s): 
        if s[i]=="-":
            flag = 1
            i+=1
        elif s[i]=="+":
            i+=1
    
    res=0
    while (i<len(s) and s[i] in numberset):
        res = res*10 + ord(s[i]) - ord("0")
        i+=1

    MAX_VALUE = 2**31-1
    MIN_VALUE = -2**31
    if res:
        if flag:
            return max(MIN_VALUE,-1*res)
        else:
            return min(res,MAX_VALUE)
    else:
        return 0


s=" -3213"
print(myatoi(s))

File: java/Apple/Arrays_and_Strings/myatoi.java
package Apple.Arrays_and_Strings;
//import java.util.*
public class myatoi {
    public int atoi(String s){
        // empty test
        if (s.isEmpty()) return 0;

        // skip the space
        int i = 0;
        while(i<s.length() && s.charAt(i)==' ')
            ++i;

        
        // neg / pos 
        int sign = 1;
        if (i<s.length() && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            sign = sign - (s.charAt(i) == '-' ? 2 : 0);
            ++i;
        }

        int res=0;
        while (i<s.length() && s.charAt(i)>='0' && s.charAt(i)<='9'){
            if (res==Integer.MAX_VALUE/10 && s.charAt(i)>'7') 
                {   
                    if (sign == 1)return Integer.MAX_VALUE;
                    else return Integer.MIN_VALUE;
                }
            res = res * 10 + s.charAt(i) - '0';
            ++i;
        }
        return sign*res;
    }

    public static void main(String[] args){
        myatoi sol = new myatoi();
        String a = "  ";
        System.out.println(sol.atoi(a));
    }
}